File: unmastered/ctci_problems/stacks_and_queues/3.5.js


function sortStack(stack){ //Do it with only 1 additional stack

}





































/* Space linear, time is quadratic
Set stack2ActualLength to 0
set stack2CurrLenght to 0
while stack1 is empty
	1) pop Each item off of stack 1 and push onto stack2, keeping track of the min and incrementing stack2CurrLength each time
	2) while stack2CurrLength is > stack2ActualLength
		pop stack2 off and decrement stack2ActualLength.  If min !== undefined && popped !== min, push onto stack1.
		else if min exists then set toAddBackToStack2 to the popped value and set min to undefined
	3) Push toPushBackToStack2 onto stack2 and increment stack2ActualLength
pop everything off of stack 2 and push onto stack 1.
*/

File: unmastered/ctci_problems/stacks_and_queues/3.5.py
"""
naive:
keep track of min and max 
"""

def sortStack(stackA):
	stackB = [];
	pushBackToACount = 0;

	while (len(stackA) > 0):
		pushToB = stackA.pop()
		while (len(stackB) > 0 and stackB[len(stackB) - 1] > pushToB):
			stackA.append(stackB.pop());
			pushBackToACount += 1;
		stackB.append(pushToB);
		while (pushBackToACount > 0):
			stackB.append(stackA.pop());
			pushBackToACount -= 1;

	while (len(stackB) > 0):
		stackA.append(stackB.pop());

def printStack(stack):
	print(stack);

def test():
	stackA = [5, 10, 7, 16, 0, 18];
	sortStack(stackA);
	printStack(stackA);

test();


